include/boost/url/impl/url.ipp

```diff
@@ -2046,22 +2047,92 @@ url&
 url::
 normalize()
 {
-#if 0
-    // scheme
-    {
-        char* it = s_
-            + offset(id_scheme);
-        auto last = it + offset(id_user);
-        if(it < last)
+    char* dest = s_;
+    char buf{'\0'};
+    int all_ids[] = {id_scheme, id_user, id_pass,
+                      id_host,  id_port, id_path,
+                      id_query, id_frag};
+    int next_id_offset = offset(id_scheme);
+    for (auto id: all_ids)
+    {
+        char* it = s_ + next_id_offset;
+        char* end = s_ + offset(id + 1);
+        while (it < end)
+        {
+            if (*it != '%')
+            {
+                *dest = *it;
+                ++it;
+                ++dest;
+                continue;
+            }
+            if (end - it < 3)
+                break;
+
+            // decode unreserved octets
+            pct_decode_unchecked(
+                &buf,
+                &buf + 1,
+                string_view(it, 3));
+            if (unreserved_chars(buf))
+            {
+                *dest = buf;
+                it += 3;
+                ++dest;
+                continue;
+            }
+
+            // uppercase percent-encoding triplets
+            ++it;
+            *it = grammar::ascii_toupper(*it);
+            ++it;
+            *it = grammar::ascii_toupper(*it);
+            ++it;
+            dest += 3;
+        }
+        // update offset of next component
+        next_id_offset = offset_[id + 1];
+        std::size_t new_offset = dest - s_;
+        std::size_t diff =
+            offset_[id + 1] - new_offset;
+        offset_[id + 1] -= diff;
+    }
+    *dest = '\0';
+    offset_[id_end] = dest - s_;
+
+    // scheme and host to lowercase
```

Comment:
These need to be in their own functions `normalize_scheme` and `normalize_host`

Reply:
Yep. I'll create `normalize_*` functions then.